crypto/bio/bss_dgram.c

```diff
@@ -790,6 +976,715 @@ static int dgram_puts(BIO *bp, const char *str)
     return ret;
 }
 
+# if M_METHOD == M_METHOD_WSARECVMSG
+static void translate_msg_win(BIO *b, WSAMSG *mh, WSABUF *iov,
+                              unsigned char *control, BIO_MSG *msg)
+{
+    iov->len = msg->data_len;
+    iov->buf = msg->data;
+
+    /* Windows requires namelen to be set exactly */
+    mh->name = msg->peer != NULL ? &msg[0].peer->sa : NULL;
+    if (msg->peer != NULL && dgram_get_sock_family(b) == AF_INET)
+        mh->namelen = sizeof(struct sockaddr_in);
+#  if OPENSSL_USE_IPV6
+    else if (msg->peer != NULL && dgram_get_sock_family(b) == AF_INET6)
+        mh->namelen = sizeof(struct sockaddr_in6);
+#  endif
+    else
+        mh->namelen = 0;
+
+    /*
+     * When local address reception (IP_PKTINFO, etc.) is enabled, on Windows
+     * this causes WSARecvMsg to fail if the control buffer is too small to hold
+     * the structure, or if no control buffer is passed. So we need to give it
+     * the control buffer even if we aren't actually going to examine the
+     * result.
+     */
+    mh->namelen         = msg->peer != NULL ? (ULONG)sizeof(*msg[0].peer) : 0;
```

Comment:
I am slightly confused by this. We've gone through an elaborate procedure on lines 987-995 above to set `mh->namelen`. Then we ignore what we came up with and overwrite it here. What am I missing?

Reply:
Hm, good catch.